### PR TITLE
19038 - unit tests for short form amalgamations

### PR DESCRIPTION
**Pull request by ketaki-deodhar**
*Issue #:* /bcgov/entity#19038

*Description of changes:*

Unit tests for short form amalgamations

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/Dashboard/FilingHistoryList/FilingTemplate.vue [hunk 1]
```diff
@@ -53,7 +53,7 @@
       </div>
     </v-expansion-panel-header>
 
-    <v-expansion-panel-content class="pt-5">
+    <v-expansion-panel-content eager class="pt-5">
```
Comment (ketaki-deodhar) [inline src/components/Dashboard/FilingHistoryList/FilingTemplate.vue:56]: Based on discussion thread below added `eager` to the `v-expansion-panel-content`. This way panel content is generated.
https://github.com/vuejs/vue-test-utils/issues/1827

File: tests/unit/AmalgamationFiling.spec.ts [hunk 2]
```diff
@@ -49,13 +50,122 @@ describe('Amalgamation Filing', () => {
...
     // verify content
     expect(wrapper.find('.item-header-title').text()).toBe('Amalgamation Application - Regular')
     expect(wrapper.find('.item-header-subtitle').text()).toContain('FILED AND PAID')
     expect(wrapper.find('.item-header-subtitle').text()).toContain('(filed by John Doe on Jan 1, 2021)')
     expect(wrapper.find('.item-header-subtitle').text()).toContain('EFFECTIVE as of Jan 1, 2021')
+    expect(wrapper.find('.view-details').text()).toBe('View Documents')
+    expect(wrapper.find('.hide-details').text()).toBe('Hide Documents')
 
-    // FUTURE: expand the panel and verify content
+    await wrapper.find('.view-details').trigger('click')
```
Comment (ketaki-deodhar) [inline tests/unit/AmalgamationFiling.spec.ts:62]: Testing content after `v-expansion-panel` is expanded

File: tests/unit/LedgerConsentContinuationOut.spec.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ describe('Display ConsentContinuationOut component', () => {
     expect(wrapper.find('.consent-continuation-out').text()).toContain('FILED AND PAID  (filed by  on [unknown])')
     expect(wrapper.find('.consent-continuation-out').text()).toContain('EFFECTIVE as of [unknown]')
     expect(wrapper.find('.expand-btn').text()).toContain('View Documents')
-    expect(wrapper.find('.v-expansion-panel-content').exists()).toBe(false)
+    expect(wrapper.find('.v-expansion-panel-content').exists()).toBe(true)
```
Comment (ketaki-deodhar) [inline tests/unit/LedgerConsentContinuationOut.spec.ts:39]: Fixed it after adding

File: tests/unit/LedgerContinuationOut.spec.ts [hunk 4]
```diff
@@ -42,7 +42,7 @@ describe('Display ContinuationOut component', () => {
     expect(wrapper.find('.continuation-out').text()).toContain('FILED AND PAID  (filed by  on [unknown])')
     expect(wrapper.find('.continuation-out').text()).toContain('EFFECTIVE as of [unknown]')
     expect(wrapper.find('.expand-btn').text()).toContain('View Documents')
-    expect(wrapper.find('.v-expansion-panel-content').exists()).toBe(false)
+    expect(wrapper.find('.v-expansion-panel-content').exists()).toBe(true)
```
Comment (ketaki-deodhar) [inline tests/unit/LedgerContinuationOut.spec.ts:45]: Fixed it after adding `eager` to the `v-expansion-panel-content`

File: src/components/Dashboard/FilingHistoryList/FilingTemplate.vue [hunk 5]
```diff
@@ -53,7 +53,10 @@
       </div>
     </v-expansion-panel-header>
 
-    <v-expansion-panel-content class="pt-5">
+    <v-expansion-panel-content
+      eager
+      class="pt-5"
+    >
```
Comment (ketaki-deodhar) [inline src/components/Dashboard/FilingHistoryList/FilingTemplate.vue:59]: Based on discussion thread below added `eager` to the `v-expansion-panel-content`. This way panel content is generated and can be tested in the unit test.
https://github.com/vuejs/vue-test-utils/issues/1827

Tested a bit to make sure functionality and user experience is same.

Comment (severinbeauvais) [reply src/components/Dashboard/FilingHistoryList/FilingTemplate.vue:59]: This is very useful! I know some previous unit tests were either disabled or not completed because expansion panels did not seem to open in the unit tests. Thanks for the research and sharing 👍 

PS Consider sharing this at the next Developer Demo.

File: tests/unit/AmalgamationFiling.spec.ts [hunk 6]
```diff
@@ -16,12 +16,13 @@ const vuetify = new Vuetify({})
...
 const businessStore = useBusinessStore()
 
-describe('Amalgamation Filing', () => {
+describe('Regular amalgamation Filing', () => {
   let wrapper: Wrapper<AmalgamationFiling>
 
   beforeAll(() => {
     // init store
     businessStore.setLegalName('MY COMPANY')
+    sessionStorage.setItem('TEMP_REG_NUMBER', null)
```
Comment (severinbeauvais) [inline tests/unit/AmalgamationFiling.spec.ts:25]: Why is this needed?

Conversely, why is setting BUSINESS_ID in session storage not needed?

File: tests/unit/AmalgamationFiling.spec.ts [hunk 7]
```diff
@@ -49,13 +50,122 @@ describe('Amalgamation Filing', () => {
...
+    expect(wrapper.find('.completed-ia-details').exists()).toBe(true)
+    expect(wrapper.find('h4').text()).toBe('Amalgamation Complete')
+    expect(wrapper.find('p').text()).toBe('MY COMPANY has been successfully amalgamated.')
+
+    wrapper.destroy()
+  })
+})
+
+describe('Horizontal amalgamation Filing', () => {
+  let wrapper: Wrapper<AmalgamationFiling>
```
Comment (severinbeauvais) [inline tests/unit/AmalgamationFiling.spec.ts:72]: We don't always use a type for the wrapper but it's a good idea so that `wrapper.vm` contains the correct properties/methods 👍

File: tests/unit/App.spec.ts [hunk 8]
```diff
@@ -2546,3 +2546,410 @@ describe('App as a completed amalgamation application', () => {
...
+          },
+          amalgamationApplication: {
+            nameRequest: {
+              legalType: CorpTypeCd.BC_COMPANY
+            },
+            contactPoint: {
+              email: 'tester@test.com',
+              phone: '(123) 456-7890'
+            },
+            type: AmalgamationTypes.HORIZONTAL
```
Comment (severinbeauvais) [inline tests/unit/App.spec.ts:2600]: 👍

File: tests/unit/EntityHeader.spec.ts [hunk 9]
```diff
@@ -273,6 +273,172 @@ describe('Entity Header - data', () => {
...
+      router,
+      propsData: { businessId: 'BC1234567', tempRegNumber: null }
+    })
+    await Vue.nextTick()
+
+    expect(wrapper.find('#entity-legal-name').text()).toBe('Numbered Limited Company')
+    expect(wrapper.find('#business-description').text()).toBe('BC Limited Company')
+  })
+
+  it('displays draft numbered amalgamation application - Vertical', async () => {
```
Comment (severinbeauvais) [inline tests/unit/EntityHeader.spec.ts:360]: Again, there doesn't seem to be a difference between horizontal and vertical (except for the display name). What do you think about combining/reducing these tests, maybe with a comment saying they're the same for H and V?

File: tests/unit/TodoList1.spec.ts [hunk 10]
```diff
@@ -1684,6 +1684,166 @@ describe('TodoList - UI - Amalgamation Applications', () => {
...
+    rootStore.tasks = [
+      {
+        task: {
+          filing: {
+            header: {
+              name: FilingTypes.AMALGAMATION_APPLICATION,
+              status: FilingStatus.DRAFT
+            },
+            amalgamationApplication: {
+              type: AmalgamationTypes.HORIZONTAL
```
Comment (severinbeauvais) [inline tests/unit/TodoList1.spec.ts:1700]: 👍

File: tests/unit/FilingHistoryList2.spec.ts [hunk 11]
```diff
@@ -145,6 +150,44 @@ filings.forEach((filing: any, index: number) => {
...
+      expect(item.status).toBe(FilingStatus.COMPLETED)
+
+      // verify display
+      expect(wrapper.find('.item-header-title').text()).toBe('Amalgamation Application - Vertical')
+      expect(wrapper.find('.item-header-subtitle').text()).toContain('FILED AND PAID')
+      expect(wrapper.find('.item-header-subtitle').text()).toContain('(filed by Registry Staff on Feb 3, 2023)')
+      expect(wrapper.find('.item-header-subtitle').text()).toContain('EFFECTIVE as of Feb 3, 2023')
+      expect(wrapper.find('.view-details').text()).toBe('View Documents')
+      expect(wrapper.find('.hide-details').text()).toBe('Hide Documents')
+    })
```
Comment (severinbeauvais) [inline tests/unit/FilingHistoryList2.spec.ts:189]: Very nice! This goes along with updating the sample filings response (filings.json below).

Now you know everything :)

**Review by severinbeauvais — APPROVED** [review]
Looks great. Doesn't _need_ any updates but I'm wondering about combining the tests that don't differ in data (horiz vs vert).

**Review by PaulGarewal — APPROVED** [review]
LGTM 👍 Great learning opportunity for these tests 😃

**Review by jamespaologarcia — APPROVED** [review]
LGTM

File: tests/unit/AmalgamationFiling.spec.ts [hunk 12]
```diff
@@ -49,13 +50,122 @@ describe('Amalgamation Filing', () => {
...
+      vuetify,
+      propsData: {
+        filing: {
+          comments: [],
+          commentsCount: 0,
+          commentsLink: null,
+          displayName: 'Amalgamation Application - Horizontal',
+          documents: [],
+          documentsLink: 'dummy_link',
+          data: {},
```
[the remark this answers is not in the document]
Comment (ketaki-deodhar) [reply tests/unit/AmalgamationFiling.spec.ts:88]: I don't think this needs amalgamation type as it only asserts `displayName`. 
The only difference is the `displayName` for H and V but can be reduced to 1 with a comment

For ref:
Will be similar for Hand V
image

File: tests/unit/EntityHeader.spec.ts (hunk 9 above)
Comment (ketaki-deodhar) [reply tests/unit/EntityHeader.spec.ts:360]: Same as above I will reduce these with a comment.

File: tests/unit/AmalgamationFiling.spec.ts (hunk 6 above)
Comment (ketaki-deodhar) [reply tests/unit/AmalgamationFiling.spec.ts:25]: Somehow, the test was not able to proceed because it was failing at `v-if="!!tempRegNumber && isStatusCompleted"`.
I tried to set few attributes but only this worked.

image

Can you suggest what else I could try and it might work?

Comment (severinbeauvais) [reply tests/unit/AmalgamationFiling.spec.ts:25]: Well, I guess the emulated session storage in the tests contained left-over data for that value.

There are 2 approaches to having expected data in store/session:
1. at the end of other tests, clean up
2. at the beginning of your test, initialize as needed

I think option 2 is preferable since it doesn't depend on the other tests. So your solution is fine (but you could add a code comment like, "ensure tempRegNumber doesn't exist").

**Comment by severinbeauvais** [comment]
Looks great. Ready to merge it?

**Comment by ketaki-deodhar** [comment]
> Looks great. Ready to merge it?

Yes. Merging now
